### PR TITLE
Refactor expected type determination

### PR DESCRIPTION
**Pull request by mmcgr**
This PR introduces an arguable cleaner approach to determining type. This has the direct benefit of allowing determination of expected value types when no values are expected. 

Before:
```
expected = Dict(:a => Int64[])
->
"/:output/:a" => []
```
After
```
expected = Dict(:a => Int64[])
->
"/:output/:a/Int64" => []
```

### CODE REVIEW

File: src/code-util.jl [hunk 1]
```diff
@@ -124,24 +124,49 @@ function is_special_symbol(symbol::Symbol)::Bool
...
+    for e_type in fieldtypes(type)
+        result *= type_string(e_type)
     end
-    return type_string(input[1])
+
+    return result
 end
 
 function type_string(input::Tuple)
+    type = typeof(input)
```
Comment (mcmcgrath13) [inline src/code-util.jl:146]:
```suggestion
function type_string(input::T) where { T <: Tuple}
    type = T
```
but also maybe just use `T` directly?

File: src/code-util.jl [hunk 2]
```diff
@@ -124,24 +124,49 @@ function is_special_symbol(symbol::Symbol)::Bool
...
+    for e_type in fieldtypes(type)
+        result *= type_string(e_type)
     end
+
     return result
 end
 
-function type_string(input)
-    return "/" * string(typeof(input))
+type_string(input) = "/" * string(typeof(input))
```
Comment (mcmcgrath13) [inline src/code-util.jl:155]:
```suggestion
type_string(input::T) where T = "/" * string(T)
```

File: src/code-util.jl [hunk 3]
```diff
@@ -124,24 +124,49 @@ function is_special_symbol(symbol::Symbol)::Bool
...
+
+function type_string(type::DataType)
```
Comment (mcmcgrath13) [inline src/code-util.jl:157]: `DataType` is a bit odd to see in an annotation, would something like `(::Type{T}) where T` work?

Comment (mmcgr) [reply src/code-util.jl]: It does work, and done.

File: src/code-util.jl [hunk 4]
```diff
@@ -123,25 +123,55 @@ function is_special_symbol(symbol::Symbol)::Bool
...
 end
 
-# Generate a string representing the Rel type for the input
-# Expected inputs are a vector of types, a tuple of types, or a type
-function type_string(input::Vector)
-    if isempty(input)
-        return ""
+# Generate a string representing the Rel type for input Vectors
+# :a => [...]
+function type_string(input::Vector{T}) where T
```
Comment (mcmcgrath13) [inline src/code-util.jl:128]: is it possible for this to be a `Union` type? I think with the way we do specializations, the answer is no - is that your understanding as well?

File: src/code-util.jl [hunk 5]
```diff
@@ -123,25 +123,55 @@ function is_special_symbol(symbol::Symbol)::Bool
...
-    return "/" * string(typeof(input))
+# Generate a string representing the Rel type for input values
+# :a => 1
+type_string(input::T) where T = "/" * string(T)
+
+# Generate a string representing the Rel type for input types
+# We've extracted the type from the input so now we can convert it to a String result
+function type_string(type::Type{T}) where T
+    # Strings (otherwise treated as containers by Julia)
+    type == String && return "/$type"
```
Comment (mcmcgrath13) [inline src/code-util.jl:164]: this method looks half updated (`test` vs `T`), but the tests pass - could we add some test coverage here?

Comment (mcmcgrath13) [reply src/code-util.jl:164]: ah, wait - I see it now. `type == T` - could we just use `T` here?

**Review by mcmcgrath13 — APPROVED** [review]
thanks!

File: src/code-util.jl (hunk 4 above)
Comment (mmcgr) [reply src/code-util.jl]: I wasn't sure what you were thinking of Union-ing, but it inspired me to simplify the methods.

File: src/code-util.jl (hunk 5 above)
Comment (mmcgr) [reply src/code-util.jl]: We can and now we have. I've simplified this further, so much that I think it needs re-approval to check it still makes sense.

**Pull request merged**